text/3700-ptr-assume-moved.md

```diff
@@ -0,0 +1,257 @@
+(in the sense of an 'address' being an index into the memory array).  From the LLVM perspective, the
+high bits are part of the address and changing them means we are now dealing with a different
+address altogether.  Having to reconcile those two views necessarily creates some friction and extra
+considerations.
+
+Function signature, documentation and implementation:
+
+```
+/// Assume that the object pointed to by a pointer has been moved to a new address
+///
+/// Intended for use with pointer tagging architecture features such as AArch64 TBI.
+/// This function creates a new pointer with the address `new_address` and a brand new provenance,
+/// then assumes that a move from the original address to the new address has taken place.
+/// Note that this is only an indication for the compiler - nothing actually gets moved or reallocated.
+///
+/// SAFETY: Users *must* ensure that `new_address` actually contains the same memory as the original.
+/// The primary use-case is working with various architecture pointer tagging schemes, where two
+/// different 64-bit addresses can point to the same chunk of memory due to some bits being ignored.
+/// When used incorrectly, this function can be used to violate the memory model in arbitrary ways.
+/// Furthermore, after using this function, users must ensure that the underlying memory is only ever
+/// accessed through the newly created pointer. Any accesses through the original pointer
+/// (or any pointers derived from it) would be Undefined Behaviour.
+/// Additionally, this function is only designed for use with heap allocations. Trying to use it with
+/// a pointer to a stack-allocated variable will result in Undefined Behaviour and
+/// should not be done under any circumstances.
+#[inline(never)]
+#[unstable(feature = "ptr_assume_moved", issue = "none")]
+#[cfg_attr(not(bootstrap), rustc_simulate_allocator)]
+#[allow(fuzzy_provenance_casts)]
+pub unsafe fn assume_moved<T>(original: *mut T, new_address: usize) -> *mut T {
+    // FIXME(strict_provenance_magic): I am magic and should be a compiler intrinsic.
+    let mut ptr = new_address as *mut T;
+    // SAFETY: This does not do anything
+    unsafe {
+        asm!(
+            "/* simulate a move from {original} to {ptr} */",
+            original = in(reg) original,
+            ptr = inout(reg) ptr
+        );
+    }
+    // FIXME: call Miri hooks to update the address of the original allocation
+    ptr
+}
+```
+
+To ensure that the function actually has the desired behaviour, we need to make sure that it is
+treated similarly to real allocator functions in the codegen stage. That is to say, the function
+return value must be annotated with `noalias` in LLVM, as explained earlier in this section. One way
+to do so would be through a rustc built-in attribute similar to e.g. `rustc_allocator` -
+`rustc_simulate_allocator`. This attribute will be passed down to the codegen stage so that the
+codegen can appropriately annotate the function.
```

Comment:
So how does this interact with allocation-eliding optimizations.

Reply:
I expect it doesn't have any allocation-eliding optimizations in LLVM, since those require more than just `noalias` on the return type, e.g. `malloc` has the additional function attributes:
`allockind("alloc,uninitialized") allocsize(0) "alloc-family"="malloc"`

Reply:
I'm not sure what exactly you mean. `noalias` in LLVM has different semantics when used with function return values. From the LLVM documentation:

> On function return values, the noalias attribute indicates that the function acts like a system memory allocation function, returning a pointer to allocated storage disjoint from the storage for any other object accessible to the caller.